File: service/lib/agama/storage/config_conversions/to_json_conversions/logical_volume.rb
# frozen_string_literal: true

# Copyright (c) [2024] SUSE LLC
#
# All Rights Reserved.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of version 2 of the GNU General Public License as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
# more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, contact SUSE LLC.
#
# To contact SUSE LLC about this file by physical or electronic mail, you may
# find current contact information at www.suse.com.

require "agama/storage/config_conversions/to_json_conversions/base"
require "agama/storage/config_conversions/to_json_conversions/with_encryption"
require "agama/storage/config_conversions/to_json_conversions/with_filesystem"
require "agama/storage/config_conversions/to_json_conversions/with_size"
require "agama/storage/configs/logical_volume"

module Agama
  module Storage
    module ConfigConversions
      module ToJSONConversions
        # Logical volume conversion to JSON hash according to schema.
        class LogicalVolume < Base
          include WithEncryption
          include WithFilesystem
          include WithSize

          # @see Base
          def self.config_type
            Configs::LogicalVolume
          end

        private

          # @see Base#conversions
          def conversions
            {
              index:      config.index,
              alias:      config.alias,
              encryption: convert_encryption,
              filesystem: convert_filesystem,
              size:       convert_size,
              name:       config.name,
              stripes:    config.stripes,
              stripeSize: config.stripe_size&.to_i,
              pool:       config.pool?,
              usedPool:   config.used_pool
            }
          end
        end
      end
    end
  end
end

File: web/src/storage/helpers/logical-volume.ts
/*
 * Copyright (c) [2025] SUSE LLC
 *
 * All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, contact SUSE LLC.
 *
 * To contact SUSE LLC about this file by physical or electronic mail, you may
 * find current contact information at www.suse.com.
 */

import { model } from "~/api/storage";
import { copyApiModel, buildLogicalVolume } from "~/storage/helpers/api-model";
import { data } from "~/storage";

function findVolumeGroupIndex(apiModel: model.Config, vgName: string): number {
  return (apiModel.volumeGroups || []).findIndex((v) => v.vgName === vgName);
}

function findLogicalVolumeIndex(volumeGroup: model.VolumeGroup, mountPath: string): number {
  return (volumeGroup.logicalVolumes || []).findIndex((l) => l.mountPath === mountPath);
}

function addLogicalVolume(
  apiModel: model.Config,
  vgName: string,
  data: data.LogicalVolume,
): model.Config {
  apiModel = copyApiModel(apiModel);

  const vgIndex = findVolumeGroupIndex(apiModel, vgName);
  if (vgIndex === -1) return apiModel;

  const volumeGroup = apiModel.volumeGroups[vgIndex];
  const logicalVolume = buildLogicalVolume(data);

  volumeGroup.logicalVolumes ||= [];
  volumeGroup.logicalVolumes.push(logicalVolume);

  return apiModel;
}

function editLogicalVolume(
  apiModel: model.Config,
  vgName: string,
  mountPath: string,
  data: data.LogicalVolume,
): model.Config {
  apiModel = copyApiModel(apiModel);

  const vgIndex = findVolumeGroupIndex(apiModel, vgName);
  if (vgIndex === -1) return apiModel;

  const volumeGroup = apiModel.volumeGroups[vgIndex];

  const lvIndex = findLogicalVolumeIndex(volumeGroup, mountPath);
  if (lvIndex === -1) return apiModel;

  const oldLogicalVolume = volumeGroup.logicalVolumes[lvIndex];
  const newLogicalVolume = { ...oldLogicalVolume, ...buildLogicalVolume(data) };

  volumeGroup.logicalVolumes.splice(lvIndex, 1, newLogicalVolume);

  return apiModel;
}

function deleteLogicalVolume(
  apiModel: model.Config,
  vgName: string,
  mountPath: string,
): model.Config {
  apiModel = copyApiModel(apiModel);

  const vgIndex = findVolumeGroupIndex(apiModel, vgName);
  if (vgIndex === -1) return apiModel;

  const volumeGroup = apiModel.volumeGroups[vgIndex];

  const lvIndex = findLogicalVolumeIndex(volumeGroup, mountPath);
  if (lvIndex === -1) return apiModel;

  volumeGroup.logicalVolumes.splice(lvIndex, 1);

  return apiModel;
}

export { addLogicalVolume, editLogicalVolume, deleteLogicalVolume };
